rfcs/0181-list-index-syntax.md

```diff
@@ -0,0 +1,290 @@
+---
+feature: list-index-syntax
+start-date: 2024-07-14
+author: rhendric
+co-authors:
+shepherd-team: infinisil inclyc
+shepherd-leader:
+related-issues: https://github.com/NixOS/nix/issues/10949, https://github.com/NixOS/rfcs/pull/137
+---
+
+# Summary
+[summary]: #summary
+
+This proposal extends the attrpath syntax to include `'[' INT ']'` components that refer to elements in lists.
+This would enable expressions such as the following:
+
+```nix
+x.[0]       # = builtins.elemAt x 0
+x.[1] or y  # = if builtins.isList x && builtins.length x > 1 then builtins.elemAt x 1 else y
```

Comment:
I believe it should abort evaluation unless it is an indexable.

Reply:
In the current language, `or` doesn't cause an abort if the selection target isn't a set:

```nix
nix-repl> null.x or false
false
```

Why should this proposal follow different rules?

Reply:
I was hit by a precedent, then forget this.